### DISCUSSION TITLE
Support wildcards in UpstreamHost as Nginx does

### DISCUSSION THREAD

**Discussion by davidni** [ThreeMammals/Ocelot#1818] [Ideas]
## New Feature
Support routes that match wildcard sub-domains such as `*.domain.example.com` (and fix existing bugs in the process -- more on this below :))

## Motivation for New Feature
`ReRoute.UpstreamHost` uses ordinal string comparison and does not support wildcards.

## Implementation note
For reference, [here's what Nginx supports](http://nginx.org/en/docs/http/server_names.html) (look under the `Wildcard names` section in particular). My suggestion that seems to align with Nginx ideas:
1. Parse the host header to extract the host and optional port. There is currently a bug in Ocelot where if the Host header is `Host: example.com:123` it will not match a route with UpstreamHost `example.com`, which IMHO it obviously should
2. If there's a trailing dot in the incoming host name, remove it (i.e. `example.com` and `example.com.` should be treated the same)
3. `UpstreamHost = "*.example.com"` would be matched with regex `^[a-z0-9.-]+\.example\.com$`
4. `UpstreamHost = ".example.com"` would be matched with regex `^(?:[a-z0-9.-]+\.)?example\.com$`

(4) is nice because it is fairly common for a service to want to take over a domain and all of its subdomains, and it offers an easy way to do that with a single route.

I'll start working on this and will send a PR soon (maybe tomorrow). Let me know if you have any concerns or suggestions.

**Comment by TomPallister** [comment]
@davidni awesome this all sounds good to me, with PR lets just make sure there are tests for all the functionality you are going to add :D If you can be bothered updating the docs with all of this would also be awesome but I can do that if you dont want to! Really appreciate your contributions.

**Comment by davidni** [comment]
@TomPallister quick update since it's been a while. I haven't given up on this issue, but priorities changed a bit and I haven't been able to get to it. I still want to get to it asap, but there's other stuff on the way.

As a side note, I'm very happy with the 10.* release and that so many dependencies are gone. Makes it a lot easier for us to consume! Thanks and keep it up.

**Comment by TomPallister** [comment]
@davidni all good! I'm going to take out everything from Ocelot apart from FluentValidation and see where things are. I want to try and make the whole thing more extensible so people can spin up their own packages to get stuff done / maybe override Ocelot's default behaviour.

**Comment by rafaelcaviquioli** [comment]
I need wildcards to work with multiple staging environments

**Reply by raman-m** [reply]
Please show us your user scenario, not saying "I need". Any practical user scenarios with good description are welcome.

**Comment by raman-m** [comment]
@davidni commented [on Jul 19, 2018](#discussion-5906959):
> I'll start working on this and will send a PR soon (maybe tomorrow). Let me know if you have any concerns or suggestions.

Hello Mr. David!
I would be happy if you let us know your plans on delivery a PR for review!
Do you have an intention to contribute?

Pay attention you've discussed 8th version with Tom, and nowadays we have 19th version for .NET 7.
A bit different product, right?
I guess your wildcard idea for upstream hosts is still actual.

---

@TomPallister 
This is so nice! 🤣 5 years! 🎉 
I believe the issue shows good idea. But user scenario is still unclear for me. I don't understand why we need to merge all subdomains into one super route definition. Would not it be risky?
It is dilemma:
- we have one super route which processes all subdomains traffic
- we will split and register multiple subdomains routes using dynamic routes feature in Service Discovery area

What do you think?

**Comment by raman-m** [comment]
> ### Implementation note
> For reference, [here's what Nginx supports](http://nginx.org/en/docs/http/server_names.html) (look under the Wildcard names section in particular).

Why not to use just Nginx proxy with this functionality? 😉 

-----
> My suggestion that seems to align with Nginx ideas:

Ah, got it... You had just idea... and nothing more! 🤣

**Comment by AlyHKafoury** [comment]
@raman-m Can I work on it ?

**Reply by raman-m** [reply]
@AlyHKafoury  Dear Aly,
You should understand that [nginx | Server names](https://nginx.org/en/docs/http/server_names.html) feature and Nginx has different philosophy than Ocelot one. Nginx functions as network traffic proxy (aka firewall with reverse proxy function), but Ocelot makes routing requests with deep protocol object transforms based on routes, for example in case of `http(s)` protos a `HttpRequest` obj is transformed deeply based on Route settings.
I'm worrying about mixing both philosophies.

**Reply by AlyHKafoury** [reply]
@raman-m I am sure you have more experience than me working with Ocelot and if that is your opinion, I can contribute to other areas that are more helpful to the project and the user-base.

**Reply by raman-m** [reply]
@AlyHKafoury Shall we continue our conversation here?

**Comment by zxyhardware-sys** [comment]
hi there ,does the new version support wildcards on matching upstream host? i've googled but seems not. here is my scenario why i need this:
i have two website, one is a company official website which supports multilingal with the language code as the subdomain, like "en.xxx.com","cn.xxx.com" etc..., and the other web is a multinational-tenant system and it supports user-custom domain by setting CNAME. but for some reason (lack of money), i need to host these two web in one server, and i want to use ocelot to route all "*.xxx.com" to the first web , and all others user-custom domains will be routed to the second web.

i have googled how but seems ocelot don't support this?thx.

**Reply by raman-m** [reply]
Will you help implement all these functionalities? Read the [Contributing](https://github.com/ThreeMammals/Ocelot?tab=readme-ov-file#contributing) in ReadMe and [Contributing](https://ocelot.readthedocs.io/en/latest/releasenotes.html#contributing) doc first please!
